### PR TITLE
Implement compile-time boundary checks around inline arrays access.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -8083,7 +8083,27 @@ BoundExpression bindInlineArrayElementAccess(ExpressionSyntax node, BoundExpress
                     }
                 }
 
-                // PROTOTYPE(InlineArrays): Check bounds
+                // Check bounds
+                if (convertedIndex.ConstantValueOpt is { SpecialType: SpecialType.System_Int32, Int32Value: int constIndex })
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8087]: Does this cover integer constants of other types, e.g., `Int16`?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:8087]: > Does this cover integer constants of other types, e.g., Int16?

We do not need to cover scenarios like that here. We can end up here only with three types: int, System.Index and System.Range.  See ``tryImplicitConversionToInlineArrayIndex````, for example.

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 2]
```diff
@@ -12326,5 +12326,408 @@ .locals init (System.Span<int> V_0)
...
+        _ = f[new System.Index(11)];
+
+        _ = f[new System.Index(12, true)];
+        _ = f[new System.Index(11, true)];
+        _ = f[new System.Index(10, true)];
+        _ = f[new System.Index(1, true)];
+        _ = f[new System.Index(0, true)];
+        _ = f[new System.Index(-1, true)];
+        _ = f[new System.Index(int.MinValue, true)];
+    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:12468]: >} [](http://example.com/codeflow?start=4&length=1)

Consider testing:
```
f[new System.Index()];
f[new System.Index(value: -1, fromEnd: false)];
f(new System.Index(fromEnd: false, value: -1)];
```

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:12473]: > Consider testing: ...

Added.

**Review by cston — APPROVED** [review]

**Pull request merged**